### PR TITLE
Support Xamarin.Forms

### PR DESCRIPTION
**Pull request by wcoder**
### Description of Change ###

Milestone: https://github.com/Softeq/XToolkit.WhiteLabel/milestone/4?closed=1

### New solution projects

- WhiteLabel.Essentials - The projects that extend WhiteLabel;
- WhiteLabel.Forms - WhiteLabel for Xamarin.Forms;
- Playground.Forms - Projects for debugging Xamarin.Forms support.

### New solution entry point

```
samples/Playground.sln -> XToolkit.sln
```

### New solution structure
Screen Shot 2020-03-31 at 7 33 53 PM

### API Changes ###

Moved:
- WhiteLabel.iOS.ImagePicker/* -> WhiteLabel.Essentials.iOS.ImagePicker/*
- WhiteLabel.Android.ImagePicker/* -> WhiteLabel.Essentials.Android.ImagePicker/*
- Softeq.XToolkit.Permissions.IPermissionRequestHandler -> Softeq.XToolkit.Common.Droid.Permissions.IPermissionRequestHandler 

### Platforms Affected ### 

- Core (all platforms)
- iOS
- Android

### PR Checklist ###

- [x] Targets the correct branch
- [x] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Softeq.XToolkit.WhiteLabel.Forms/Behaviors/EventToCommandBehavior.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            }
+            if (AssociatedObject == null)
+            {
+                return;
+            }
+
+            var eventInfo = AssociatedObject.GetType().GetRuntimeEvent(name);
+            if (eventInfo == null)
+            {
+                throw new ArgumentException($"{nameof(EventToCommandBehavior)}: Can't de-register the '{EventName}' event.");
```
Comment (SavikPavel) [inline Softeq.XToolkit.WhiteLabel.Forms/Behaviors/EventToCommandBehavior.cs:113]: Same here, maybe it's better to use "unregister" instead

File: samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/ModalPage.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Developed by Softeq Development Corporation
+// http://www.softeq.com
+
+namespace Playground.Forms.Views.Dialogs.Modal
+{
+    public partial class ModalPage
```
[the remark this answers is not in the document]
Comment (wcoder) [reply samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/ModalPage.xaml.cs:6]: partial, inherited in xaml

File: samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/SecondModalPage.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+// Developed by Softeq Development Corporation
+// http://www.softeq.com
+
+namespace Playground.Forms.Views.Dialogs.Modal
+{
+    public partial class SecondModalPage
```
[the remark this answers is not in the document]
Comment (wcoder) [reply samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/SecondModalPage.xaml.cs:6]: partial, inherited in xaml

File: samples/Playground.Forms/Playground.Forms.iOS/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("Playground.Forms.iOS")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
```
[the remark this answers is not in the document]
Comment (wcoder) [reply samples/Playground.Forms/Playground.Forms.iOS/Properties/AssemblyInfo.cs:13]: You right! Created issue #284

File: samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/ModalPage.xaml.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SavikPavel) [reply samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/ModalPage.xaml.cs:6]: Don't you think that we should implicitly set inherited class here at least for the ease of reading?

File: samples/Playground/Playground.Droid/CustomDroidBootstrapper.cs [hunk 5]
```diff
@@ -37,7 +37,7 @@ protected override void ConfigureIoc(IContainerBuilder builder)
             builder.Singleton<RequestResultHandler, IPermissionRequestHandler>();
 
             // image picker
-            builder.Singleton<DroidImagePickerService, IImagePickerService>();
+            builder.Singleton<DroidImagePickerService, Softeq.XToolkit.WhiteLabel.Essentials.ImagePicker.IImagePickerService>();
```
[the remark this answers is not in the document]
Comment (wcoder) [reply samples/Playground/Playground.Droid/CustomDroidBootstrapper.cs:40]: Fixed

File: Softeq.XToolkit.WhiteLabel.Forms/Behaviors/EventToCommandBehavior.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            if (eventInfo == null)
+            {
+                throw new ArgumentException($"{nameof(EventToCommandBehavior)}: Can't register the '{EventName}' event.");
+            }
+            var methodInfo = typeof(EventToCommandBehavior).GetTypeInfo().GetDeclaredMethod(nameof(OnEvent));
+            _eventHandler = methodInfo.CreateDelegate(eventInfo.EventHandlerType, this);
+            eventInfo.AddEventHandler(AssociatedObject, _eventHandler);
+        }
+
+        private void DeregisterEvent(string name)
```
[the remark this answers is not in the document]
Comment (wcoder) [reply Softeq.XToolkit.WhiteLabel.Forms/Behaviors/EventToCommandBehavior.cs:95]: Awesome! I agree!
Fixed

File: Softeq.XToolkit.WhiteLabel.Forms/Behaviors/EventToCommandBehavior.cs (hunk 1 above)
Comment (wcoder) [reply Softeq.XToolkit.WhiteLabel.Forms/Behaviors/EventToCommandBehavior.cs:113]: Fixed

File: samples/Playground/Playground/ViewModels/Pages/DetailsPageViewModel.cs [hunk 7]
```diff
@@ -25,7 +25,9 @@ public override void OnAppearing()
             base.OnAppearing();
 
             var title = Person?.FullName ?? string.Empty;
-            var message = Person == null ? "You navigated without parameter" : "You navigated with parameter:";
+            var message = Person == null
+                ? "You navigated without parameter"
+                : $"You navigated with parameter: {Person.FullName}";
```
[the remark this answers is not in the document]
Comment (wcoder) [reply samples/Playground/Playground/ViewModels/Pages/DetailsPageViewModel.cs:30]: Changed

File: Softeq.XToolkit.WhiteLabel.Forms/FormsPlatformProvider.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        }
+
+        public Task OnUIThreadAsync(Action action)
+        {
+            return Device.InvokeOnMainThreadAsync(action);
+        }
+
+        public void OnUIThread(Action action)
+        {
+            throw new NotImplementedException("Method unsupported");
```
[the remark this answers is not in the document]
Comment (wcoder) [reply Softeq.XToolkit.WhiteLabel.Forms/FormsPlatformProvider.cs:27]: Reasonable question!
I think we need to discuss and rework all the `Execute.OnUIThread`, related to #259

File: samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/ModalPage.xaml.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wcoder) [reply samples/Playground.Forms/Playground.Forms/Views/Dialogs/Modal/ModalPage.xaml.cs:6]: I think at this point we might miss it.
Also, we need to discuss it in the future.
Maybe we want to use `PageBase<TViewModel>`

File: samples/Playground.Forms/Playground.Forms/ViewModels/Dialogs/DialogsRootPageViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            get => _dialogResult;
+            set => Set(ref _dialogResult, value);
+        }
+
+        private async Task ShowModalScreenAsync()
+        {
+            var result = await _dialogsService
+                .For<ModalPageViewModel>()
+                .WithParam(x => x.Title, "With")
```
[the remark this answers is not in the document]
Comment (wcoder) [reply samples/Playground.Forms/Playground.Forms/ViewModels/Dialogs/DialogsRootPageViewModel.cs:46]: We use the fluent interface everywhere for navigation/dialogs. This gives more flexibility. Approach with a simple method only to support old codebase.

**Pull request merged**